README.md
# py-cgmail
A module that leverages pyzmail to parse email messages into json and parses urls out of the message body

# Installation

## Ubuntu
  ```bash
  $ sudo apt-get install -y python-dev python-pip git
  $ pip install git+https://github.com/csirtgadgets/py-cgmail
  ```

# Examples
```
$ cgmail -h
$ cgmail-cif -h
```

# Support and Documentation

You can also look for information at the [GitHub repo](https://github.com/csirtgadgets/py-cgmail).

# License and Copyright

Copyright (C) 2015 [the CSIRT Gadgets Foundation](http://csirtgadgets.org)

Free use of this software is granted under the terms of the [GNU Lesser General Public License](https://www.gnu.org/licenses/lgpl.html) (LGPL v3.0). For details see the file ``LICENSE`` included with the distribution.


CHANGELOG.rst

Changelog
=========

0.1.0 (2015-08-11)
-----------------------------------------

* First release on GitHub.


tests/test_cgmail.py

from cgmail.__main__ import main


def test_main():
    assert main([]) == 0


src/cgmail/cli.py
#!/usr/bin/env python

# http://blog.magiksys.net/parsing-email-using-python-content
# https://github.com/mailgun/flanker
# https://pregmatch.org/read/python-procmail
# http://stackoverflow.com/questions/557906/want-procmail-to-run-a-custom-python-script-everytime-a-new-mail-shows-up
# http://stackoverflow.com/questions/14676375/pipe-email-from-procmail-to-python-script-that-parses-body-and-saves-as-text-fil
# http://blog.magiksys.net/parsing-email-using-python-content
# https://github.com/andris9/mailparser


from argparse import ArgumentParser
from argparse import RawDescriptionHelpFormatter
import logging
import os.path
import textwrap
import sys
import yaml
import cgmail

from cifsdk.client import Client
from cifsdk.observable import Observable

LOG_FORMAT = '%(asctime)s - %(levelname)s - %(name)s[%(lineno)s] - %(message)s'
REMOTE_DEFAULT = "http://localhost:5000"


def main():

    #
    # initialize module
    #

    p = ArgumentParser(
        description=textwrap.dedent('''\
        example usage:
            $ cat test.eml | cif-procmail -v
            $ cif-procmail --file test.eml
        '''),
        formatter_class=RawDescriptionHelpFormatter,
        prog='cif-procmail'
    )

    p.add_argument("-v", "--verbose", dest="verbose", action="count",
                   help="set verbosity level [default: %(default)s]")
    p.add_argument('-d', '--debug', dest='debug', action="store_true")

    p.add_argument("-f", "--file", dest="file", help="specify email file")

    # cif arguments
    p.add_argument("--confidence", help="specify confidence for submitting to CIF", default="65")
    p.add_argument("--remote", help="specify CIF remote [default: %(default)s",
                   default=REMOTE_DEFAULT)
    p.add_argument("--token", help="specify CIF token")
    p.add_argument("--config", help="specify CIF config [default: %(default)s",
                   default=os.path.expanduser("~/.cif.yml"))
    p.add_argument("--tags", help="specify CIF tags [default: %(default)s", default=["phishing"])
    p.add_argument("--group", help="specify CIF group [default: %(default)s", default="everyone")
    p.add_argument("--tlp", help="specify CIF TLP [default: %(default)s", default="amber")
    p.add_argument("--no-verify-ssl", action="store_true", default=False)
    p.add_argument("--raw", help="include raw message data")
    p.add_argument("--raw-headers", help="include raw header data", action="store_true")
    p.add_argument("--provider", help="specify feed provider [default: %(default)s", default="localhost")

    args = p.parse_args()

    loglevel = logging.WARNING
    if args.verbose:
        loglevel = logging.INFO
    if args.debug:
        loglevel = logging.DEBUG

    console = logging.StreamHandler()
    logging.getLogger('').setLevel(loglevel)
    console.setFormatter(logging.Formatter(LOG_FORMAT))
    logging.getLogger('').addHandler(console)
    logger = logging.getLogger(__name__)

    options = vars(args)

    if os.path.isfile(args.config):
        f = file(args.config)
        config = yaml.load(f)
        f.close()
        if not config['client']:
            raise Exception("Unable to read " + args.config + " config file")
        config = config['client']
        for k in config:
            if not options.get(k):
                options[k] = config[k]

        if config.get("remote") and options["remote"] == REMOTE_DEFAULT:
            options["remote"] = config["remote"]

    else:
        logger.info("{} config does not exist".format(args.config))

    #
    # get email from file or stdin
    #

    if not options.get("remote"):
        logger.critical("missing --remote")
        raise SystemExit

    if options.get("file"):
        with open(options["file"]) as f:
            email = f.read()
    else:
        email = sys.stdin.read()

    #
    # parse email message
    #

    # parse email into message and message parts
    message, message_parts = cgmail.parse_message(email)

    # get message body from message & mail parts from message parts
    message_body = cgmail.parse_message_body(message)
    mail_parts = cgmail.parse_message_parts(message_parts)

    # extract urls from message body and mail parts
    urls = cgmail.extract_urls(message_body, mail_parts)

    #
    # submit urls to a CIF instance
    #

    # initialize cif client
    cli = Client(remote=options["remote"], token=options["token"], no_verify_ssl=options["no_verify_ssl"])

    for u in urls:
        logger.info("submitting: {0}".format(u))

        o = Observable(
            observable=u,
            confidence=options["confidence"],
            tlp=options["tlp"],
            group=options["group"],
            tags=options["tags"],
            provider=options["provider"]
        )
        r = cli.submit(submit=str(o))
        logger.info("submitted: {0}".format(r))


if __name__ == "__main__":
    main()


src/cgmail/__init__.py
# -*- encoding: utf-8 -*-
__version__ = "0.1.0"


from pyzmail.parse import message_from_string as pyzmail_message_from_string
from pyzmail.parse import get_mail_parts as pyzmail_get_mail_parts
from pyzmail.parse import decode_text as pyzmail_decode_text

try:
    import re2 as re
except ImportError:
    import re

# http://stackoverflow.com/questions/499345/regular-expression-to-extract-url-from-an-html-link
# RE for URL extraction:
# http://daringfireball.net/2010/07/improved_regex_for_matching_urls
# http://daringfireball.net/misc/2010/07/url-matching-regex-test-data.text
# https://gist.github.com/uogbuji/705383

RE_URL_PLAIN = 'http[s]?://(?:[a-zA-Z]|[0-9]|[$-_@.&+]|[!*\(\),]|(?:%[0-9a-fA-F][0-9a-fA-F]))+'


def parse_message(email):
    message = pyzmail_message_from_string(email)
    parts = pyzmail_get_mail_parts(message)
    return message, parts


def parse_message_body(msg):
    # Do we need to process text/plain and text/html separately? 
    # It's not clear to me that we do. 
    body = None
    if msg.get_content_type() == "text/plain":
        body = msg.get_payload(decode=True)
        if body:
            if msg.get_charset():
                decoded_body = pyzmail_decode_text(body, msg.get_charset(), None)
                body = decoded_body[0]
            elif msg.get_charsets():
                for c in msg.get_charsets():
                    decoded_body = pyzmail_decode_text(body, c, None)
                    body = decoded_body[0]
            else:
                decoded_body = pyzmail_decode_text(body, msg.get_charset(), None)
                body = decoded_body[0]
    elif msg.get_content_type() == "text/html":
        body = msg.get_payload(decode=True)
        if body:
            if msg.get_charset():
                decoded_body = pyzmail_decode_text(body, msg.get_charset(), None)
                body = decoded_body[0]
            elif msg.get_charsets():
                for c in msg.get_charsets():
                    decoded_body = pyzmail_decode_text(body, c, None)
                    body = decoded_body[0]
            else:
                decoded_body = pyzmail_decode_text(body, msg.get_charset(), None)
                body = decoded_body[0]
    return body


def parse_message_headers(msg):
    msg_headers = {}
    for header in msg.keys():
        header = header.lower()
        value = msg.get_decoded_header(header)
        try:
            msg_headers[header].append(value)
        except KeyError:
            msg_headers[header] = [value]
    return msg_headers


def parse_message_parts(message_parts):
    mail_parts = []
    for p in message_parts:
        d = {}
        d["charset"] = p.charset
        d["content_id"] = p.content_id
        d["description"] = p.description
        d["filename"] = p.filename
        d["is_body"] = p.is_body
        d["sanitized_filename"] = p.sanitized_filename
        d["type"] = p.type

        if p.charset:
            try:
                d["payload"] = p.get_payload().decode(p.charset)
            except UnicodeDecodeError:
                decoded_body = pyzmail_decode_text(p.get_payload(), None, None)
                d["payload"] = decoded_body[0]
        else:
            decoded_body = pyzmail_decode_text(p.get_payload(), None, None)
            d["payload"] = decoded_body[0]
        mail_parts.append(d)
    return mail_parts


def extract_urls(message_body, mail_parts):

    links = set()

    if message_body:
        urls = re.findall(RE_URL_PLAIN, message_body)

        for u in urls:
            links.add(u)

    elif mail_parts:
        for p in mail_parts:
            # if the mail_part is large skip the regex search
            # 500,000 is a number pulled out of this air
            if int(len(p['payload']) > 500000):
                continue
            if p['payload']:
                urls = re.findall(RE_URL_PLAIN, p['payload'])

                for u in urls:
                    links.add(u)
    else:
        return links

    return links


src/cgmail/__main__.py
# -*- encoding: utf-8 -*-

# Why does this file exist, and why __main__?
# For more info, read:
# - https://www.python.org/dev/peps/pep-0338/
# - https://docs.python.org/2/using/cmdline.html#cmdoption-m
# - https://docs.python.org/3/using/cmdline.html#cmdoption-m
# - https://chriswarrick.com/blog/2014/09/15/python-apps-the-right-way-entry_points-and-scripts/

import sys
import cgmail
import json
import logging
import os.path
import textwrap
import yaml

from argparse import ArgumentParser
from argparse import RawDescriptionHelpFormatter
from cifsdk.client import Client
from cifsdk.observable import Observable

LOG_FORMAT = '%(asctime)s - %(levelname)s - %(name)s[%(lineno)s] - %(message)s'
REMOTE_DEFAULT = "http://localhost:5000"


def print_json(message_headers, message_body, mail_parts, urls):

    # turn set into list to serialize in json
    list_urls = list(urls)

    # create dictionary of data structures
    d = {
        'headers': message_headers,
        'message_body': message_body,
        'mail_parts': mail_parts,
        'urls': list_urls
    }

    # convert dictionary to json and print to screen
    print(json.dumps(d, indent=4, sort_keys=True))


def submit_to_cif(logger, options, urls):

    # initialize cif client
    cli = Client(remote=options["remote"], token=options["token"], no_verify_ssl=options["no_verify_ssl"])

    for u in urls:
        logger.info("submitting: {0}".format(u))

        o = Observable(
            observable=u,
            confidence=options["confidence"],
            tlp=options["tlp"],
            group=options["group"],
            tags=options["tags"],
            provider=options["provider"]
        )
        r = cli.submit(submit=str(o))
        logger.info("submitted: {0}".format(r))
        

def main():

    #
    # initialize module
    #

    p = ArgumentParser(
        description=textwrap.dedent('''\
        example usage:
            $ cat test.eml | cif-procmail -v
            $ cif-procmail --file test.eml
        '''),
        formatter_class=RawDescriptionHelpFormatter,
        prog='cif-procmail'
    )

    p.add_argument("-v", "--verbose", dest="verbose", action="count",
                   help="set verbosity level [default: %(default)s]")
    p.add_argument('-d', '--debug', dest='debug', action="store_true")
    p.add_argument("-f", "--file", dest="file", help="specify email file")
    p.add_argument('-p', '--print', dest='print_stdout', action='store_true',
                   default=False, help='print json to stdout')
    p.add_argument('--cif-urls', dest='cif_urls', action='store_true', help='submit extracted urls to CIF')

    # cif arguments
    p.add_argument("--confidence", help="specify confidence for submitting to CIF", default="65")
    p.add_argument("--remote", help="specify CIF remote [default: %(default)s",
                   default=REMOTE_DEFAULT)
    p.add_argument("--token", help="specify CIF token")
    p.add_argument("--config", help="specify CIF config [default: %(default)s",
                   default=os.path.expanduser("~/.cif.yml"))
    p.add_argument("--tags", help="specify CIF tags [default: %(default)s", default=["phishing"])
    p.add_argument("--group", help="specify CIF group [default: %(default)s", default="everyone")
    p.add_argument("--tlp", help="specify CIF TLP [default: %(default)s", default="amber")
    p.add_argument("--no-verify-ssl", action="store_true", default=False)
    p.add_argument("--raw", help="include raw message data")
    p.add_argument("--raw-headers", help="include raw header data", action="store_true")
    p.add_argument("--provider", help="specify feed provider [default: %(default)s", default="localhost")

    args = p.parse_args()

    loglevel = logging.WARNING
    if args.verbose:
        loglevel = logging.INFO
    if args.debug:
        loglevel = logging.DEBUG

    console = logging.StreamHandler()
    logging.getLogger('').setLevel(loglevel)
    console.setFormatter(logging.Formatter(LOG_FORMAT))
    logging.getLogger('').addHandler(console)
    logger = logging.getLogger(__name__)

    options = vars(args)

    if os.path.isfile(args.config):
        f = file(args.config)
        config = yaml.load(f)
        f.close()
        if not config['client']:
            raise Exception("Unable to read " + args.config + " config file")
        config = config['client']
        for k in config:
            if not options.get(k):
                options[k] = config[k]

        if config.get("remote") and options["remote"] == REMOTE_DEFAULT:
            options["remote"] = config["remote"]

    else:
        logger.info("{} config does not exist".format(args.config))

    #
    # get email from file or stdin
    #

    if not options.get("remote"):
        logger.critical("missing --remote")
        raise SystemExit

    if options.get("file"):
        with open(options["file"]) as f:
            email = f.read()
    else:
        email = sys.stdin.read()

    #
    # parse email message
    #

    # parse email into message and message parts
    message, message_parts = cgmail.parse_message(email)

    # get message headers, body and mail parts
    message_body = cgmail.parse_message_body(message)
    message_headers = cgmail.parse_message_headers(message)
    mail_parts = cgmail.parse_message_parts(message_parts)

    # extract urls from message body and mail parts
    urls = cgmail.extract_urls(message_body, mail_parts)

    if args.print_stdout:
        print_json(message_headers, message_body, mail_parts, urls)

    #
    # submit urls to a CIF instance
    #

    if args.cif_urls:
        submit_to_cif(logger, options, urls)

if __name__ == "__main__":
    main()
